dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java

```diff
@@ -0,0 +1,249 @@
+    }
+
+    @Override
+    public List<VanityUrl> getAllVanityUrls(final User user) {
+        ImmutableList.Builder<VanityUrl> results = ImmutableList.builder();
+        try {
+            List<Contentlet> contentResults = contentletAPI
+                    .search("+baseType:" + BaseContentType.VANITY_URL.getType() + " +working:true",
+                            0, 0, StringPool.BLANK, user, false);
+            contentResults.stream().forEach((Contentlet con) -> {
+                results.add(getVanityUrlFromContentlet(con));
+            });
+        } catch (DotDataException | DotSecurityException e) {
+            Logger.error(this, "Error searching vanityurl URLs", e);
+        }
+        return results.build();
+    }
+
+    @Override
+    public VanityUrl getWorkingVanityUrl(final String uri, final Host host, final long languageId,
+            final User user) {
+        return getVanityUrlByURI(uri, host, languageId, user, false);
+    }
+
+    @Override
+    public VanityUrl getLiveVanityUrl(final String uri, final Host host, final long languageId,
+            final User user) {
+        return getVanityUrlByURI(uri, host, languageId, user, true);
+    }
+
+    protected VanityUrl getVanityUrlByURI(final String uri, final Host host, final long languageId,
+            final User user, final boolean live) {
+        VanityUrl result = vanityURLCache.get(VanityUrlUtil.sanitizeKey(
+                host != null && InodeUtils.isSet(host.getInode()) ? host.getIdentifier()
+                        : Host.SYSTEM_HOST, VanityUrlUtil.fixURI(uri), languageId));
+        if (result == null || !InodeUtils.isSet(result.getInode())) {
+            List<VanityUrl> results = new ArrayList<>();
+            String hostId = (host != null ? host.getIdentifier() : Host.SYSTEM_HOST);
+            try {
+                List<Contentlet> contentResults = contentletAPI
+                        .search("+baseType:" + BaseContentType.VANITY_URL.getType()
+                                + " +languageId:" + languageId + " +conHost:" + hostId
+                                + " +vanityUrl:" + VanityUrlUtil.fixURI(uri) + (live ? " +live:true"
+                                : " +working:true"), 0, 0, StringPool.BLANK, user, false);
+                contentResults.stream().forEach((Contentlet con) -> {
+                    VanityUrl vanityUrl = getVanityUrlFromContentlet(con);
+                    try {
+                        if (con.isLive()) {
+                            addToVanityURLCache(vanityUrl);
+                        } else {
+                            VanityUrlServices.invalidateVanityUrl(vanityUrl);
+                        }
+                    } catch (DotDataException | DotSecurityException e) {
+                        Logger.error(this, "Error processing Vanity Url - contentlet Id:" + con
+                                .getIdentifier(), e);
+                    }
+                    results.add(vanityUrl);
```

Comment:
Almost the entire code from 84 to 101 is then repeated from 111 to 129. Please extract to a method with a clear name and description of what it does